times_excel_reader.py

```diff
@@ -81,148 +124,250 @@ def has_tag(cls, tag):
     )
-    df.drop(index=comment_rows, inplace=True)
-    df.reset_index(drop=True, inplace=True)
+    if len(missing) > 0:
+        print(f"WARNING: Missing {len(missing)} tables: {missing_str}")
+
+    total_gt_rows = 0
+    total_correct_rows = 0
+    for table_name, gt_table in sorted(
+        ground_truth.items(), reverse=True, key=lambda t: len(t[1].values)
+    ):
+        total_gt_rows += len(gt_table.values)
+        if table_name in data:
+            data_table = data[table_name]
+
+            # Remove .integer suffix added to duplicate column names by CSV reader (mangle_dupe_cols=False not supported)
+            transformed_gt_cols = [col.split(".")[0] for col in gt_table.columns]
+
+            if transformed_gt_cols != list(data[table_name].columns):
+                print(
+                    f"WARNING: Table {table_name} header incorrect, was"
+                    f" {data_table.columns.values}, should be {transformed_gt_cols}"
+                )
+            else:
+                # both are in string form so can be compared without any issues
+                gt_rows = set(tuple(row) for row in gt_table.values.tolist())
+                data_rows = set(tuple(row) for row in data_table.values.tolist())
+                total_correct_rows += len(gt_rows.intersection(data_rows))
+                additional = data_rows - gt_rows
+                missing = gt_rows - data_rows
+                if len(additional) != 0 or len(missing) != 0:
+                    print(
+                        f"WARNING: Table {table_name} ({data_table.shape[0]} rows,"
+                        f" {gt_table.shape[0]} GT rows) contains {len(additional)}"
+                        f" additional rows and is missing {len(missing)} rows"
+                    )
+                    DataFrame(additional).to_csv(
+                        os.path.join("output", table_name + "_additional.csv"),
+                        index=False,
+                    )
+                    DataFrame(missing).to_csv(
+                        os.path.join("output", table_name + "_missing.csv"), index=False
+                    )
+
+    print(
+        f"{total_correct_rows / total_gt_rows :.1%} of ground truth rows present"
+        f" in output ({total_correct_rows}/{total_gt_rows})"
+    )
+
+
+#############################################################################
+################             TRANSFORM FUNCTIONS             ################
+#############################################################################
+
+
+def remove_comment_rows(table: EmbeddedXlTable) -> EmbeddedXlTable:
+    """
+    Return a modified copy of 'table' where rows with cells containig '*'
+    or '\I:' in their first of third columns have been deleted. These characters
+    are defined in https://iea-etsap.org/docs/Documentation_for_the_TIMES_Model-Part-IV.pdf
+    as comment identifiers.
+    TODO: is this description correct?
```

Comment:
Yes

Reply:
Cool, thanks. Why only first and third columns though? What's the reasoning behind it? It would be good add it to the comment.

Reply:
I think that is a bug.  Only the first column should count.

Reply:
Hm, interestingly, if I run the program without that block of code (so we only delete the first column) the program fails with:
`Traceback (most recent call last):
  File "times_excel_reader.py", line 2122, in <module>
    tables = convert_xl_to_times(xl_files_dir, input_files, mappings)
  File "times_excel_reader.py", line 251, in convert_xl_to_times
    output = transform(input)
  File "times_excel_reader.py", line 817, in process_user_constraint_tables
    return [process_user_constraint_table(t) for t in tables]
  File "times_excel_reader.py", line 817, in <listcomp>
    return [process_user_constraint_table(t) for t in tables]
  File "times_excel_reader.py", line 813, in process_user_constraint_table
    df["Year"] = df["Year"].astype("Int64")
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/generic.py", line 6245, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 446, in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/internals/managers.py", line 348, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/internals/blocks.py", line 527, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/dtypes/astype.py", line 299, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/dtypes/astype.py", line 230, in astype_array
    values = astype_nansafe(values, dtype, copy=copy)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/dtypes/astype.py", line 95, in astype_nansafe
    return dtype.construct_array_type()._from_sequence(arr, dtype=dtype, copy=copy)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/arrays/masked.py", line 132, in _from_sequence
    values, mask = cls._coerce_to_array(scalars, dtype=dtype, copy=copy)
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/arrays/numeric.py", line 258, in _coerce_to_array
    values, mask, _, _ = _coerce_to_data_and_mask(
  File "/home/developer/hackathon/times-excel-reader2/.venv/lib/python3.8/site-packages/pandas/core/arrays/numeric.py", line 212, in _coerce_to_data_and_mask
    values = values.astype(dtype, copy=copy)
ValueError: invalid literal for int() with base 10: '2021-2050'`

Reply:
So maybe there's a third column somewhere that actually has some comments and this was just a quick way to get rid of it? Not sure how to check which table this is though.

Reply:
For me, Visual Studio opens the debugger on that line and I can inspect the variables to determine the table.  The row in question is not a comment, we are just failing to parse it.

Reply:
I see. In the interest of getting this merged and avoiding conflicts due to code changes to the underlying code I'm going to leave this as it is, adding a TODO comment explaining it. Then anyone can look into this at any point.